Make copies of returned reference types stored in widget state

AnOctopus opened the pull request:
#3600  is caused by the initial value of a widget, which is stored and
compared to internally, being returned directly as the widget's return
value. This is not a problem for value types, but if we return a mutable
reference type and the user mutates it later in the script, it changes
our stored copy, so we incorrectly think the initial value has changed
on the next run.

This is also a potential issue for values that are passed in, in various
ways, but handling them systematically will require some careful
checking. The reference type sharing introduces a potential for bugs
akin to "shotgun parsing", and Python is not very well equiped for
dealing with this kind of issue.

The unit test for this might make more sense elsewhere, but I couldn't
decide where.

AnOctopus committed:
Make copies of returned reference types stored in widget state

#3600 is caused by the initial value of a widget, which is stored and
compared to internally, being returned directly as the widget's return
value. This is not a problem for value types, but if we return a mutable
reference type and the user mutates it later in the script, it changes
our stored copy, so we incorrectly think the initial value has changed
on the next run.

This is also a potential issue for values that are passed in, in various
ways, but handling them systematically will require some careful
checking. The reference type sharing introduces a potential for bugs
akin to "shotgun parsing", and Python is not very well equiped for
dealing with this kind of issue.

The unit test for this might make more sense elsewhere, but I couldn't
decide where.

willhuang1997 commented on lib/tests/streamlit/state/session_state_test.py:
```diff
@@ -233,6 +233,23 @@ def test_widget_presence(self):
+        color.append("purple")
+
+        ctx = get_report_ctx()
+        assert ctx.session_state._initial_widget_values["color"] is not color
```
For this assertion, maybe im missing something but is checking that they're not the same a good test?
I am thinking that checking that we should check that purple got added?

AnOctopus replied:
The purpose of this assertion is to verify that `ctx.session_state._initial_widget_values["color"]` does _not_ get purple added. The test would perhaps be simpler if I left off the purple part, because I used `is` rather than `==` so that it would check reference identity rather than value equality, so the purple append is unnecessary

AnOctopus committed:
Remove redundant append from state aliasing test

The test already uses `is not` to check reference identity instead of
value equality, so the append is unnecessary and misleading.

AnOctopus committed:
Give UploadedFile metadata value equality

This is to make the file uploader e2e tests pass, because they relied on
reference equality of UploadedFile just working out. It isn't strictly
clear that doing this is correct, since someone could somehow rely on
the previous behavior of only being equal when they pointed at the same
object in memory, but UploadedFile is a small wrapper class for BytesIO
with some metadata attributes, so I don't expect people to treat it as
more than an opaque blob or possibly extract some of those attributes
very often, and issues would involve mutating the returned
UploadedFile (since two instances that were previously equal will still
be equal, and this only is misleading if just one is mutated). So it
seems safe to do this.

AnOctopus committed:
Revert UploadedFile equality change

It breaks a unit test, and I haven't investigated why it exists enough
to decide the change should go through, so instead I will fix the e2e
test by applying the change only within it.
